src/sources/kafka.rs

```diff
@@ -370,105 +411,559 @@ impl SourceConfig for KafkaSourceConfig {
+/// when partitions are assigned. When a shutdown signal is received, or
+/// partitions are being revoked, the Consuming state is traded for a Draining
+/// state (and associated drain deadline future) via the `begin_drain` method
+///
+/// A ConsumerStateInner<Draining> keeps track of partitions that are expected
+/// to complete, and also owns the signal that, when dropped, indicates to the
+/// client driver task that it is safe to proceed with the rebalance or shutdown.
+/// When draining is complete, or the deadline is reached, Draining is traded in for
+/// either a Consuming (after a revoke) or Complete (in the case of shutdown) state,
+/// via the `finish_drain` method.
+///
+/// A ConsumerStateInner<Complete> is the final state, reached after a shutdown
+/// signal is received. This can not be traded for another state, and the
+/// coordination task should exit when this state is reached.
+struct ConsumerStateInner<S> {
+    config: KafkaSourceConfig,
+    decoder: Decoder,
+    out: SourceSender,
+    log_namespace: LogNamespace,
+    consumer_state: S,
+}
+struct Consuming;
+struct Complete;
+struct Draining {
+    signal: SyncSender<()>,
+
+    /// The set of topic-partition tasks that are required to complete during
+    /// the draining phase, populated at the beginning of a rebalance or shutdown.
+    /// Partitions that are being revoked, but not being actively consumed
+    /// (e.g. due to the consumer task exiting early) should not be included.
+    /// The draining phase is considered complete when this set is empty.
+    expect_drain: HashSet<TopicPartition>,
+
+    /// Whether the client is shutting down after draining. If set to true,
+    /// the `finish_drain` method will return a Complete state, otherwise
+    /// a Consuming state.
+    shutdown: bool,
+}
+type OptionDeadline = OptionFuture<Pin<Box<Sleep>>>;
+enum ConsumerState {
+    Consuming(ConsumerStateInner<Consuming>),
+    Draining(ConsumerStateInner<Draining>),
+    Complete(ConsumerStateInner<Complete>),
+}
+impl Draining {
+    fn new(signal: SyncSender<()>, shutdown: bool) -> Self {
+        Self {
+            signal,
+            shutdown,
+            expect_drain: HashSet::new(),
+        }
     }
 
-    let mut stream = consumer.stream();
+    fn is_complete(&self) -> bool {
+        self.expect_drain.is_empty()
+    }
+}
 
-    loop {
-        tokio::select! {
-            biased;
-            _ = &mut shutdown => break,
-            entry = ack_stream.next() => if let Some((status, entry)) = entry {
-                if status == BatchStatus::Delivered {
-                    if let Err(error) =
-                        consumer.store_offset(&entry.topic, entry.partition, entry.offset)
-                    {
-                        emit!(KafkaOffsetUpdateError { error });
+impl<C> ConsumerStateInner<C> {
+    pub fn complete(self, _deadline: OptionDeadline) -> (OptionDeadline, ConsumerState) {
```

Comment:
I'm not clear what the point of this parameter is if it's always thrown away.

Also, a lot of these methods are marked as `pub`, but I don't see them used outside of this module. Was there a problem with not having the `pub` marker?